### PR TITLE
updated implementation to comply with S3 Read/Write API endpoint spec.

### PR DESCRIPTION
**Pull request by JuliaFernee**

### CODE REVIEW

**Comment by nwrigh** [comment]
Ignore the semaphoreci thing, it was a 10% day thing that I didn't realise hooked in

File: service/handlers_test.go [hunk 1]
```diff
@@ -86,11 +88,47 @@ func TestAddAdminHandlers(t *testing.T) {
 	})
 }
 
+func TestMethodHandlerAcceptsNonEmptyResourcePath(t *testing.T) {
+	r := mux.NewRouter()
```
Comment (david-smith-ft) [inline service/handlers_test.go]: Having re-read these tests, I think that they're testing the same as the other tests unnecessarily.  I think going down the named routes route would actually be better to test the setting of the resource path.
`r.Get("mh").URL("uuid", <uuid>) == /notempty/<uuid>`

Comment (JuliaFernee) [inline service/handlers_test.go]: I don't mind changing but I am worried whether there will be a quorum? Other reviewers may think otherwise...

Comment (nwrigh) [inline service/handlers_test.go]: You could take out the additional tests and see if coverage goes down if you aren't convinced. Also Pete just told me that you can see inline the coverage if you use Atom as an editor - I am going to check it out now :)

File: service/handlers.go [hunk 2]
```diff
@@ -102,8 +103,11 @@ func Handlers(servicesRouter *mux.Router, wh WriterHandler, rh ReaderHandler) {
...
-	servicesRouter.Handle("/__count", ch)
-	servicesRouter.Handle("/__ids", ih)
-	servicesRouter.Handle("/", ah)
+	if resourcePath != "" {
+		resourcePath = fmt.Sprintf( "/%s", resourcePath)
+	}
+	servicesRouter.Handle(fmt.Sprintf("%s%s", resourcePath, "/{uuid:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}}"), mh)
+	servicesRouter.Handle(fmt.Sprintf("%s%s", resourcePath, "/__count"), ch)
+	servicesRouter.Handle(fmt.Sprintf("%s%s", resourcePath, "/__ids"), ih)
+	servicesRouter.Handle(fmt.Sprintf("%s%s", resourcePath, "/"), ah)
```
Comment (carlosroman) [inline service/handlers.go]: Missing test for `/__count`, `/__ids` and `/`.

File: service/processor_test.go [hunk 3]
```diff
@@ -134,11 +136,32 @@ func TestWritingToS3(t *testing.T) {
...
 
+func TestWritingToS3WithTransactionID(t *testing.T) {
+	w, s := getWriter()
+	p := []byte("PAYLOAD")
+	ct := expectedContentType
+	var err error
+	err = w.Write(expectedUUID, &p, ct, expectedTransactionId)
+	assert.NoError(t, err)
+	assert.NotEmpty(t, s.putObjectInput)
+	assert.Equal(t, expectedTransactionId, *s.putObjectInput.Metadata[transactionid.TransactionIDKey])
```
Comment (david-smith-ft) [inline service/processor_test.go]: Why can't this assert just be added into `TestWritingToS3`? This is just duplicating that test.

File: service/processor.go [hunk 4]
```diff
@@ -490,3 +490,35 @@ func (rh *ReaderHandler) HandleGet(rw http.ResponseWriter, r *http.Request) {
...
+}
+
+func writerServiceUnavailable(uuid string, err error, rw http.ResponseWriter) {
+	log.Errorf("Error writing '%v': %v", uuid, err.Error())
+	respondServiceUnavailable(err, rw)
+}
+
+func readerServiceUnavailable(uuid string, err error, rw http.ResponseWriter) {
+	if uuid != "" {
+		log.Errorf("Error from reader: '%v': %v", uuid, err.Error())
```
Comment (JuliaFernee) [inline service/processor.go]: Sure

File: service/processor.go [hunk 5]
```diff
@@ -327,16 +329,22 @@ func (w *S3Writer) Write(uuid string, b *[]byte, ct string) error {
...
 
 	if err != nil {
-		log.Infof("Error found, Resp was : %v", resp)
+		log.Errorf("Error found, Resp was : %v", resp)
 		return err
 	}
-
 	return nil
 }
+func addTransactionID(params *s3.PutObjectInput, tid string) {
```
Comment (JuliaFernee) [inline service/processor.go]: I can change it but your suggestion incourages bad coding practices as a lot of our go code does. 
This has to do with excessive in-lining and dealing with more than one responsibility in a function/method, all of which increases its cyclomatic complexity.  Some say the function should fit in your palm, others - it should be no more that the screen length. There is a lot of literature about it - "Code Complete", Object Calystenics, articles on refactoring by ThoughtWorks etc

File: service/processor_test.go (hunk 3 above)
Comment (JuliaFernee) [inline service/processor_test.go]: They shouldn't be the same. These tests test specifically that that transaction ID is passed to S3. I have re-written them in a better way.  TestWritingToS3 is already testing more than it should. There used to be a rule that you have one assertion per unit test, but again with our "new Go coding standards" this is no longer the case.

File: service/processor.go [hunk 6]
```diff
@@ -490,3 +490,35 @@ func (rh *ReaderHandler) HandleGet(rw http.ResponseWriter, r *http.Request) {
...
+func writerServiceUnavailable(uuid string, err error, rw http.ResponseWriter) {
+	log.Errorf("Error writing '%v': %v", uuid, err.Error())
+	respondServiceUnavailable(err, rw)
+}
+
+func readerServiceUnavailable(uuid string, err error, rw http.ResponseWriter) {
+	if uuid != "" {
+		log.Errorf("Error from reader: '%v': %v", uuid, err.Error())
+	} else {
+		log.Errorf("Error from reader: %v", err.Error())
```
Comment (JuliaFernee) [inline service/processor.go]: agreed to change uuid to Request URI when logging errors

**Pull request merged**